File: psx-web/src/lib/api/backtest.ts
const API_URL = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

async function apiFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${API_URL}${path}`, {
    credentials: "include",
    headers: { "Content-Type": "application/json", ...init?.headers },
    ...init,
  });
  if (!res.ok) {
    const err = (await res.json().catch(() => ({ detail: res.statusText }))) as {
      detail?: string;
    };
    throw new Error(err.detail ?? "Request failed");
  }
  return res.json() as Promise<T>;
}

export interface StrategyMeta {
  key: string;
  label: string;
  default_params: Record<string, number>;
}

export interface BacktestRequest {
  symbol: string;
  strategy: string;
  start_date: string; // YYYY-MM-DD
  end_date: string;
  initial_capital_pkr: string;
  commission_per_side_pct?: number;
  parameters?: Record<string, number>;
}

export interface TradeOut {
  entry_date: string;
  entry_price: number;
  exit_date: string;
  exit_price: number;
  quantity: number;
  commission_pkr: number;
  gross_pnl_pkr: number;
  net_pnl_pkr: number;
  return_pct: number;
  holding_days: number;
  reason_open: string;
  reason_close: string;
  narration: string;
}

export interface EquityPointOut {
  date: string;
  equity_pkr: number;
  drawdown_pct: number;
  in_position: boolean;
}

export interface BacktestResult {
  symbol: string;
  strategy: string;
  start_date: string;
  end_date: string;
  initial_capital_pkr: number;
  final_value_pkr: number;
  total_return_pct: number;
  cagr_pct: number;
  sharpe_ratio: number;
  max_drawdown_pct: number;
  num_trades: number;
  win_rate_pct: number;
  avg_winning_trade_pct: number;
  avg_losing_trade_pct: number;
  avg_holding_days: number;
  total_commissions_pkr: number;
  benchmark_return_pct: number;
  excess_return_pct: number;
  trades: TradeOut[];
  equity_curve: EquityPointOut[];
  lessons: string[];
}

export const fetchBacktestStrategies = () =>
  apiFetch<StrategyMeta[]>("/api/v1/backtest/strategies");

export const runBacktest = (body: BacktestRequest) =>
  apiFetch<BacktestResult>("/api/v1/backtest", {
    method: "POST",
    body: JSON.stringify(body),
  });

File: psx-api/psx_api/schemas/backtest.py
from datetime import date
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, Field

StrategyKey = Literal[
    "sma_crossover",
    "rsi_oversold",
    "macd_signal",
    "bollinger_revert",
    "buy_and_hold",
]


class StrategyMeta(BaseModel):
    key: str
    label: str
    default_params: dict[str, Any]


class BacktestRequest(BaseModel):
    symbol: str = Field(..., min_length=1, max_length=20)
    strategy: StrategyKey
    start_date: date
    end_date: date
    initial_capital_pkr: Decimal = Field(default=Decimal("100000"), gt=0)
    commission_per_side_pct: float = Field(default=0.15, ge=0, le=5)
    parameters: dict[str, Any] = Field(default_factory=dict)


class TradeOut(BaseModel):
    entry_date: date
    entry_price: float
    exit_date: date
    exit_price: float
    quantity: float
    commission_pkr: float
    gross_pnl_pkr: float
    net_pnl_pkr: float
    return_pct: float
    holding_days: int
    reason_open: str
    reason_close: str
    narration: str


class EquityPointOut(BaseModel):
    date: date
    equity_pkr: float
    drawdown_pct: float
    in_position: bool


class BacktestResult(BaseModel):
    symbol: str
    strategy: str
    start_date: date
    end_date: date
    initial_capital_pkr: float
    final_value_pkr: float
    total_return_pct: float
    cagr_pct: float
    sharpe_ratio: float
    max_drawdown_pct: float
    num_trades: int
    win_rate_pct: float
    avg_winning_trade_pct: float
    avg_losing_trade_pct: float
    avg_holding_days: float
    total_commissions_pkr: float
    benchmark_return_pct: float
    excess_return_pct: float
    trades: list[TradeOut]
    equity_curve: list[EquityPointOut]
    lessons: list[str]
